app/main.py

```diff
@@ -1,213 +1,53 @@
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+from app.knight_stats import KNIGHTS
+from app.knight import Knight
+
+
+def battle(cfg):
+
+    # Knights are coming to the arena
+    lancelot = Knight(cfg["lancelot"]["name"],
+                      cfg["lancelot"]["power"],
+                      cfg["lancelot"]["hp"],
+                      cfg["lancelot"]["armour"],
+                      cfg["lancelot"]["weapon"],
+                      cfg["lancelot"]["potion"])
+
+    arthur = Knight(cfg["arthur"]["name"],
+                    cfg["arthur"]["power"],
+                    cfg["arthur"]["hp"],
+                    cfg["arthur"]["armour"],
+                    cfg["arthur"]["weapon"],
+                    cfg["arthur"]["potion"])
+
+    mordred = Knight(cfg["mordred"]["name"],
+                     cfg["mordred"]["power"],
+                     cfg["mordred"]["hp"],
+                     cfg["mordred"]["armour"],
+                     cfg["mordred"]["weapon"],
+                     cfg["mordred"]["potion"])
+
+    red_knight = Knight(cfg["red_knight"]["name"],
+                        cfg["red_knight"]["power"],
+                        cfg["red_knight"]["hp"],
+                        cfg["red_knight"]["armour"],
+                        cfg["red_knight"]["weapon"],
+                        cfg["red_knight"]["potion"])
+
+    # preparation
+    lancelot.equip().drink_potion()
+    arthur.equip().drink_potion()
+    mordred.equip().drink_potion()
+    red_knight.equip().drink_potion()
```

Comment:
can you optimize this with the help of cycles or something else, please

Reply:
well, that was my view of knight's show  presentation, but i'll optimize that.